### DISCUSSION TITLE
Convert ≔ to \coloneqq from Word to LaTeX

### DISCUSSION THREAD

**Discussion by ZhuangQu** [jgm/pandoc#10558] [Ideas]
`≔` is a unicode-math character (U+2254) in [A.docx](https://github.com/user-attachments/files/18497131/A.docx). Run
```
pandoc A.docx -o B.tex
```
and then B.tex is
```
\[≔\]
```
which causes an error. The correct output is
```
\[\coloneqq\]
```
where `\coloneqq` is the LaTeX command corresponding to `≔`.

**Comment by jgm** [comment]
This should be fine if you use xelatex or lualatex when you compile the LaTeX code; have you tried that?

**Reply by ZhuangQu** [reply]
You're right. Only pdflatex throws an error. However, pandoc translates all unicode-math characters into corresponding LaTeX commands. From what I've observed, `≔` is the only exception, breaking the consistency.

**Reply by jgm** [reply]
Actually texmath (which we use for conversion) does support this, but only when `unicode-math` is enabled in the environment:

https://github.com/jgm/texmath/blob/master/src/Text/TeXMath/Unicode/ToTeX.hs#L698

Pandoc assumes only `amsmath` and `amssymb`.  This is what you want when generating LaTeX that can be processed with `pdflatex`, since the default templates only load `unicode-math` for lualatex and xelatex.

Hence I think pandoc is doing the right thing here.

**Reply by ZhuangQu** [reply]
You mean: `pdflatex` does not support `unicode-math` package, and `\coloneqq` belongs to `unicode-math`, so `pandoc` does not support `\coloneqq`. Is my understanding correct?

However, `\coloneqq` is in `mathtools` package, and `pdflatex` support `mathtools`. Actually, `mathtools` is proposed to fixes various deficiencies of `amsmath` you mentioned, see https://ctan.org/pkg/mathtools

**Reply by jgm** [reply]
Yes, correct. I did not know about `mathtools`.
Currently we have
```
\usepackage{amsmath,amssymb}
```
in the default LaTeX template.  Would you recommend adding `mathtools` to this list?  Is it compatible with `unicode-math` (which we load when pdflatex is not being used) and `mathspec` (which we load instead of `unicode-math` when the `mathspec` variable is set and xelatex is being used)?

**Reply by ZhuangQu** [reply]
Yes, I recommend adding `mathtools` to the list.

I did some experiments and found no incompatibilities between `mathtools` and `unicode-math`/`mathspec`. By searching on stackexchange, I saw a few incompatible reports on special cases. However, there are also incompatible reports between `unicode-math` and `amsmath`. `unicode-math` also [recognizes](https://ctan.org/pkg/unicode-math) that "Care should be taken when transitioning from `amsmath`".

Compatibility issues among LaTeX packages is a perennial problem, not something we can avoid. All pandoc can do is to translate commands as many as possible.

**Comment by jpcirrus** [comment]
I have been using `mathtools` with `unicode-math` in a custom template for almost 10 years without issue. This is with both `xelatex` and `lualatex` engines. `mathtools.sty` contains the following:

```latex
\RequirePackage{amsmath}[2016/11/05]
\@ifpackageloaded{unicode-math}{
  \PackageWarning{mathtools}{For best results, please load mathtools
    *before* \MessageBreak unicode-math}
}{}
```

So line 20 of `default.latex` can be changed to `\usepackage{mathtools,amssymb}`. `unicode-math` is already loaded later in `fonts.latex`, so no change required there.

Also, `mathtools` loads `calc` unconditionally, so the 2 conditional loads of the `calc` package in `common.latex` can be removed.

**Reply by jgm** [reply]
To clarify, if we load `mathtools`, it replaces `amsmath` and we don't need to load it too?

**Comment by jpcirrus** [comment]
Correct.

**Comment by jgm** [comment]
OK, so two action items here:

- [ ] Update the default.latex template as described above by @jpcirrus
- [ ] Tell pandoc to call texmath with the `unicode-math` environment, so it will use `\coloneqq` and other symbols not available otherwise

**Reply by jgm** [reply]
texmath already supports `unicode-math`, and knows nothing about `mathtools`. So my assumption was that if we tell texmath that we have `amsmath`, `amssymb`, and `unicode-math` available, we'll get output that works well with `mathtools`.  But maybe this assumption is false.

**Reply by jgm** [reply]
Sorry, I had gotten a bit confused on this. Let me be more explicit about how texmath is working.  We have a big list of TeX command / unicode character correspondences,
https://github.com/jgm/texmath/blob/master/src/Text/TeXMath/Unicode/ToTeX.hs#L129-L2885

In some cases, this list will have different options for different "environments":  e.g.,
```
  , Record {uchar = '\120831', commands = [("base","\\mathtt{9}"),("unicode-math","\\mttnine")], category = Ord, comments = "mathematical monospace digit 9"}]
```
For this code point texmath will use `\mathtt{9}` in the base configuration, but if we tell it that `unicode-math` is available, it will use `\mttnine`.

In the case that concerns us now, we have
```
  , Record {uchar = '\8788', commands = [("mathabx","\\coloneq"),("txfonts","\\coloneqq"),("unicode-math","\\coloneq")], category = Rel, comments = "= \\SetDelayed (wrisym), # := colon, equals"}
```
Here there is no alternative for the base configuration; if `unicode-math` is available, `\coloneq` (not `\coloneqq`) will be used; if `txfonts` is available, `\coloneqq` will be used.

I can see now that just adding `mathtools` won't give us the full range of `unicode-math` symbols.  And so we're back to the problem I noted at the outset.  We're producing a tex file that might be compiled with `unicode-math`, but this isn't guaranteed, so we can't really use those symbols.

**Reply by jgm** [reply]
Currently there's nothing here specific to `mathtools` (because this data derives from a list put together by Guenter Milde in 2011, probably before `mathtools`).  I suppose someone could add information to this data about symbols in `mathtools`, and then pandoc could tell texmath that we have `mathtools`.  But someone would have to do that work.

**Reply by jgm** [reply]
The Milde list is here: https://milde.users.sourceforge.net/LUCR/Math/data/unimathsymbols.txt
The `mathtools` documentation refers to another list put together by the STIX project, https://www.ams.org/STIX/bnb/stix-tbl.asc
We could think about just using that list, but unlike the Milde list it doesn't have information about the packages that provide the command. And before making any change we'd want to get a complete catalog of the differences.

**Reply by ZhuangQu** [reply]
For the same unicode-math character, you want to translate it into different commands based on different LaTeX environments. Is my understanding correct? This is too complicated, not only hard to implement, but also hard to maintain.

I recommend that you specify a "standard" environment where all unicode-math characters have a unique translation. This is much easier to implement and maintain. What is the "standard" environment like? My suggestion is to align with GitHub markdown. Markdown's formulas are expected to be consistent with LaTeX's formulas. GitHub does not allow users to specify which LaTeX package to use, so this environment is unique (standard).

In GitHub markdown, the four commands you mentioned are shown as:
![QQ_1737651630352](https://github.com/user-attachments/assets/ca266429-8bfd-426d-b8a3-5d9e8b13adbb)
As you can see, only `\mathtt{9}` and `\coloneqq` are rendered correctly, so they are standard translations.

**Reply by jgm** [reply]
texmath is designed to allow this kind of flexibility.  We could revisit that design. However, it is important that in converting *from* tex math all of these symbols can be recognized.

Pandoc has never taken advantage of this flexibility; it just always tells texmath that is has amsmath and amssymb.

I assume that GH is using MathJax for math rendering?

**Reply by ZhuangQu** [reply]
I checked the github documentation and confirmed that MathJax is used for rendering. I have an idea: add MathJax mode to texmath, and pandoc told texmath to use MathJax. [MathJax supports these command](https://docs.mathjax.org/en/latest/input/tex/macros/index.html#environments)

**Reply by jgm** [reply]
I think it could be quite problematic to assume that all of the symbols in this list are defined in rendering LaTeX.  That would be fine for HTML + MathJaX, but it would mean that users who are producing PDFs via LaTeX would need to include a large number of packages to ensure that everything works.
e.g. `\curl` requires the `physics` package and `\centernot` requires the `centernot` package and `longRightleftharpoons` requires `mwchem` and `\upzeta` requires `upgreek`.  We don't want to put all of these in our default template.

**Reply by ZhuangQu** [reply]
I agree with you. It's up to you to decide which packages to support.

**Comment by ZhuangQu** [comment]
I find that when using `unicode-math`, there is no need to translate unicode-math characters into latex commands. For example, the following code is OK:
```
\documentclass{article}
\usepackage{unicode-math}
\begin{document}
$α≔∅$
\end{document}
```
without translating `α` to `\alpha`, `≔` to `\coloneq`, `∅` to `\emptyset`.

So texmath only needs to have two modes:
1. `unicode-math` package is used, and unicode-math characters remain unchanged.
2. `unicode-math` package is not used, and unicode-math characters are translated into latex commands. As @jpcirrus said, `\usepackage{mathtools,amssymb}` and use them as the standard.

**Reply by jgm** [reply]
- texmath already has a `unicode-math` environment.
- Adding awareness of the `mathtools` environment to texmath would be desirable, but would require some work as noted above.  I added an issue https://github.com/jgm/texmath/issues/249 that outlines what would need to be done. It's a pretty significant job.  Without that step, you still won't get `\coloneqq` even if we change the latex template to include `mathtools`.
- Note also that pandoc can't tell texmath to use the `unicode-math` environment, because pandoc seeks to produce a latex file that can be processed either by pdflatex or by xelatex/lualatex.  `\ifpdftex` and the like are used in the template to achieve this, but it means that we can't tailor the symbols used to whether `unicode-math` is used.

**Reply by ZhuangQu** [reply]
Great!

**Comment by jgm** [comment]
As a short-term workaround, you could simply use a Lua filter to replace U+2254 with `\coloneqq` in Math elements.
